CST-210: Selection API Discussion (#22)
spacetelescope/mast-aladin-lite: docs/discussion/selection-api.md

**Use case 2:** A scientist is interested in a subset of sources in a catalog. Since they loaded the catalog themselves, they know that there is an `objId`  field that uniquely identifies the source. The query might look like the following: 

```
aladin.selectSources(sources=catalog_subset, match_key="objId")
```

If they instead want to use `RA` and `DEC` to identify the source, the query might look like the following:

```
aladin.selectSources(sources=catalog_subset, match_keys=["ra", "dec"])
```

**Use case 3:** A software engineer wants to create an interoperable layer between **ipyaladin** and another visualization tool, such as [jdaviz](https://jdaviz.readthedocs.io/en/stable/). They would then need to use a combination of the APIs listed in this doc to ensure that selections in **ipyaladin** are reflected in **jdaviz** and vice versa. 

# 2. ipyaladin interfaces
## 2.1. GET
#### get_selection

```
def get_selection() -> List[Table]
```

Returns a list of **Astropy** `Table` containing the currently selected catalog source(s) per catalog.
#### get_region

```
def get_region() -> Region
```

Returns an **Astropy** `Region` with the currently selected region (if applicable)

> #### get_footprint

Reviewer: ```suggestion
#### get_footprints
```

--------------------------------------------------------------------------------
#### select_region

```
def select_region(region: Region) -> None
```

This method triggers a selection by `Region` event based on the coordinates 

**Parameter** |**Type** |**Description**
--------------- |---------------------------------------------------------------------------------------------------| -------------------------------------------------------
region |[Region](https://astropy-regions.readthedocs.io/en/stable/api/regions.Region.html#regions.Region) |The region corresponding to the sky coordinates that the user would like to select sources within

---------------------------------------------------------------------------------------------------------------------------------------------------------------------------
#### select_footprints

```
def select_footprints(footprints: List[Footprint]) -> None
```

This method triggers a selection event on a user provided list of Foorprints

| **Parameter** | **Type**        | **Description**                                      |
| ------------- | --------------- | ---------------------------------------------------- |

> | footprints    | List[Footprint] | The list of footprints the user would like to select |

Reviewer: I'm wondering if string matching on STC strings will be sufficient here for identifying footprints or whether we need to adopt some sort of key matching like we did for sources.  I think it will be fairly common to have multiple identical footprints that correspond to different observations of the same part of the sky.  Often one would want to select all such footprints (and with graphical selections, one would select all whether one wanted to or not), but I'm not sure we can assume that.  It's at least worth discussing as part of discussing footprint selections in general.

On a related note, like with catalogs we will be able to have multiple layers containing footprints.  I wonder if there will be cases where we want to be specific about which layer(s) we're selecting from in a call like this.

| **Parameter** | **Type**  | **Description**                                                |
| ------------- | --------- | -------------------------------------------------------------- |
| match_func    | func(a,b) | The function used to determine if an object should be selected |

-------------------------------------------------------------------------------
#### SelectRegion

```
/**
* @param region: Region
*/
function SelectRegion(region)
```

Triggers a `Region` selection within **aladin-lite** widget. This function would need to mock of the various state transitions expected by the FiniteStateMachines used for region selection ([example](https://github.com/cds-astro/aladin-lite/blob/master/src/js/FiniteStateMachine/CircleSelect.js#L137-L165)).

| **Parameter** | **Type** | **Description**                     |
| ------------- | -------- | ----------------------------------- |
| region        | Region   | The region to select objects within |

-----------------------------------------------------------------------
#### SelectFootprints

```
/**
* @param footprints: List[Footprint]
*/

> function SelectFootprints(footprints)

Reviewer: The question above in the python API about how to identify footprints probably applies here too.

##### Questions for CDS
- Are there recommended approaches for reconciling celestial sphere regions with 2D image plane representations in **Astropy**?
- Beyond **Astropy** regions, are there other libraries or data structures that can better handle the translation between celestial and image coordinates?
- Should we implement a warning system to inform users when regions exceed a specified angular resolution, indicating potential inaccuracies?
## 4.2 Path Forward
If the approach outlined above is accepted, we are ready to submit draft pull requests for review. These drafts will implement the APIs discussed in this document, and we are prepared to make any necessary revisions based on the feedback we receive from CDS.

Looking ahead, we anticipate further significant updates to the API. To enable a smoother development and review process, we would like the schedule more frequent touch points to help align our goals.
##### Questions for CDS
- What next steps do you see moving forward with this suggestion?
- What collaboration tools do you prefer?
## 4.3 Footprint Selection

> **Region-Based Footprint Selection:**  
> When performing a region selection, are footprints considered selected only if they are entirely contained within the region, or are footprints that overlap with the region also included? Clarifying this behavior will be important for us moving forward as we implement consistent selection logic across multiple tools.
> 

Reviewer: ```suggestion
**Footprints as catalog elelments:**  
Currently when a "catalog" is loaded and it has an `s_region` column, the footprints defined by that column are rendered in Aladin Lite.  Does this hint at possible additional unification of how footprints and catalog sources are handling (i.e., using the same events and methods to select them, and optionally displaying the other columns somewhere on selection)?  Some unification seems intuitive and could help with identifying which footprints have been selected without needing additional baggage on the graphics objects themselves.

**Region-Based Footprint Selection:**  
When performing a region selection, are footprints considered selected only if they are entirely contained within the region, or are footprints that overlap with the region also included? Clarifying this behavior will be important for us moving forward as we implement consistent selection logic across multiple tools.

```

To support user workflows, we need to expose **aladin-lite**'s features programmatically. Our initial focus is on selection functionality, which is foundational for many use cases.

We examined the current selection features available in **ipyaladin** and **aladin-lite**, and prototyped an API to support:
- selecting a list of sources ([PR](https://github.com/pcuste1/ipyaladin/pull/2))
- selecting sources within a region
	- aladin-lite([PR](https://github.com/pcuste1/aladin-lite/pull/1))
	- ipyaladin ([PR](https://github.com/pcuste1/ipyaladin/pull/3))

These proofs-of-concept work by implementing the API discussed below. We also examined footprint selection, which will require further discussion to determine the best path forward. 
## 1.1. Design Principles
#### Standard Interface Pattern
For each feature exposed by **aladin-lite** and **ipyaladin** that we intend to support, the interface should offer the following APIs:
- **Listener**: Allows developers to attach callbacks to respond to relevant events.
- **Get**: Retrieves the current state or value of the feature.
- **Set (Select)**: Updates the feature\'s state programmatically. 

> - **Misc:** Features that support the other 3,

Reviewer 2: ```suggestion
- **Misc:** Features that support the other 3.
```

#### User-Defined Catalogs
All selection interfaces are designed with the assumption that catalogs are loaded into **aladin-lite** programmatically by the user within their Jupyter notebook. Catalogs added directly via the widget interface are not considered within the scope of these APIs, ensuring greater control and consistency for programmatic workflows.
#### Layer Agnostic APIs
In the initial implementation, we make the assumption that all APIs will operate across all the currently visible layers loaded by the user in the UI. This approach mirrors the existing functionality of **aladin-lite** through the existing `GetSelection` method. 
## 1.2. Feature Requirements
#### Selection Requirements
As a notebook user/scientist I want to\...
- select a source within a catalog using a programmatic API
- select a list of sources within all loaded catalogs using a programmatic API
- select a footprint(s) using a programmatic API 
- select all sources/footprints within a defined region using a programmatic API
#### Data Retrieval Requirements 
As a notebook user/scientist I want to\...
- access the list of currently selected sources using a programmatic API 

> - access the lost of currently select footprints currently selected using a programmatic API 

Reviewer 2: ```suggestion
- access the list of currently select footprints currently selected using a programmatic API 
```

##### Questions for CDS
- Are there recommended approaches for reconciling celestial sphere regions with 2D image plane representations in **Astropy**?
- Beyond **Astropy** regions, are there other libraries or data structures that can better handle the translation between celestial and image coordinates?
- Should we implement a warning system to inform users when regions exceed a specified angular resolution, indicating potential inaccuracies?
## 4.2 Path Forward
If the approach outlined above is accepted, we are ready to submit draft pull requests for review. These drafts will implement the APIs discussed in this document, and we are prepared to make any necessary revisions based on the feedback we receive from CDS.

Looking ahead, we anticipate further significant updates to the API. To enable a smoother development and review process, we would like the schedule more frequent touch points to help align our goals.
##### Questions for CDS
- What next steps do you see moving forward with this suggestion?
- What collaboration tools do you prefer?
## 4.3 Footprint Selection
**Footprints as catalog elelments:**  

> Currently when a "catalog" is loaded and it has an `s_region` column, the footprints defined by that column are rendered in Aladin Lite.  Does this hint at possible additional unification of how footprints and catalog sources are handling (i.e., using the same events and methods to select them, and optionally displaying the other columns somewhere on selection)?  Some unification seems intuitive and could help with identifying which footprints have been selected without needing additional baggage on the graphics objects themselves.

Reviewer 2: I think this could use more clarification and/or separation into multiple sentences. Maybe just paring down some of it. Lmk if you want feedback on the specific language, but this feels a bit confusing to read through and follow. 
Also, I think we should add a portion in this section specifically about the footprintClicked existing in aladin lite but not in ipyaladin.

Reviewer: How about something like
```suggestion
Currently when a "catalog" is loaded and it has an `s_region` column, the footprints defined by that column are rendered in Aladin Lite.  Seeing this in action led to the question of whether handling of footprints and catalog sources could be more unified.  

For example, footprint selection could fire the same event(s) as object selection, and could result in displaying the metadata in a a table or popup as with catalog object selection.  The catalog columns could also help in differentiating footprints when needed.
```

Reviewer 2: That sounds perfect!

  - [3.3. SELECT](#33-select)
  - [3.4. Misc](#34-misc)

# 1. Overview

We are building tools for the Roman Research Nexus platform to enable
scientists to analyze data from the Nancy Grace Roman Space Telescope.
As part of this effort, we are developing a suite of interoperable tools
within a JupyterLab environment, including integration with Aladin Lite.

To support scientific workflows, we aim to expose Aladin Lite\'s
features programmatically. Our initial focus is on selection
functionality, which is foundational for many analysis use cases.

We evaluated the current selection features available in ipyaladin, and
prototyped programmatic extensions to support:

- selecting a list of sources
  [[PR](https://github.com/pcuste1/ipyaladin/pull/2)]

- selecting sources within a region
  [[PR](https://github.com/pcuste1/ipyaladin/pull/1)].

These proofs-of-concept worked by reaching into internal components of
Aladin Lite. However, a more robust and maintainable solution requires
official API support.

This document outlines:

- our design principles

- feature requirements

- API specifications for both Aladin Lite and ipyaladin

- several motivating use cases.

> ## 1.1. Design Principles

Reviewer: This section is really helpful!

unified API that coordinates both actions under a single interface call.

#### Consistent Data Formats

To streamline development and enhance user experience, all ipyaladin
interfaces are expected to rely on Astropy-defined object formats (e.g.,
[Table](https://docs.astropy.org/en/stable/table/index.html),
[Region](https://astropy-regions.readthedocs.io/en/stable/)).
This provides a familiar and well-documented foundation for users
working within the scientific Python ecosystem.

#### User-Defined Catalogs

All selection interfaces are designed with the assumption that catalogs
are loaded into aladin-lite programmatically by the user within their
Jupyter notebook. Catalogs added directly via the widget interface are
not considered within the scope of these APIs, ensuring greater control
and consistency for programmatic workflows.

## 1.2. Feature Requirements

#### Selection Requirements

As a notebook user/scientist I want to\...

- slect a source within a catalog using a programmatic API

- select a list of sources within all loaded catalogs using a
  programmatic API

- select all sources within a defined region using a programmatic API

> - select a footprint using a programmatic API 

Reviewer: As a general comment for all the footprint API things here, we need to support selection of multiple footprints.  Whether we move more towards a model of unifying the handling of footprints and catalog objects is TBD, but conceptually I think the handling should be fairly similar.

#### Data Retrieval Requirements 

As a notebook user/scientist I want to\...

- access the list of currently selected sources using a programmatic
  API 

> - access the region of currently selected sources, if one was used to

Reviewer: I am seeing two possible interpretations of regions handling here.  Both have pros and cons.

Are we accessing a region object here, or a list of sources that were selected by a region at some point?  

The former allows a variety of robust options including editing a region, but also ties us to some usability questions about how dynamic the implied list of selected sources is as new catalogs or other selection constraints are added.  

The latter is no doubt simpler to implement since it does a one time selection/filtering of the objects when the region is defined and doesn't need to include the region in future logic.   This makes region selection more analogous to other selection constraints like selecting by keys or selection function, since they offer a one-time selection action.  I wonder if some hybrid approach could be considered that would allow us to persist one or more user-defined regions, but allow the user to apply those as selection filters when they want.  This would be very straightforward for a Python API, but offering it graphically would be more of a UX challenge.

  select the sources, using a programmatic API 

- access the footprint currently selected using a programmatic API 

#### Listener Requirements 

As a developer, I want to\...

- listen for changes to the selected sources 

- listen for changes to the selection region

- listen for changes to the selected footprint

#### Misc. Requirements

As a notebook user/scientist I want to\...

- be able to remove the current selection using a programmatic API

> ## 1.3. Example Use Cases

Author: Add an example use case for the region selection method and listener

**Use case 1:** A scientist is interested in all sources in a catalog
with a redshift within a defined range. The query might look like the
following
```
aladin.selectSources(match_func: lambda source: source.rshift > 1 and
source.rshift < 2)
```
**Use case 2:** A scientist is interested in a subset of the sources in
a catalog. Since they loaded the catalog themselves, they know that
there is an objId  field that uniquely identifies the source. The query
might look like the following 
```
aladin.selectSources(sources=catalog_subset, match_key="objId")
```
maybe they instead want to use RA and DEC to identify the source instead
```
aladin.selectSources(sources=catalog_subset, match_keys=["ra",
"dec"])
```
**Use case 3:** A software engineer wants to create an interoperable
layer between an aladin-lite widget and another visualization tool, such
as [jdaviz](https://jdaviz.readthedocs.io/en/stable/). They would then
need to use a combination of the APIs listed in this doc to ensure that
selection in aladin-lite are reflected in jdaviz and vice versa. 

# 2. ipyaladin interfaces

## 2.1. GET

#### GetSelection
```
def GetSelection() -> Table
```

> Returns an Astropy table containing the currently selected catalog

Reviewer: I think we should spell out in a little more detail how we handle the case of multiple catalog (or footprint) layers.  The current behavior in Aladin Lite seems intuitive in that a selection action applies to all layers (at least visible ones, so visibility may be another complication for the programmatic API).  Then GetSelections returns a list-like structure (one item per layer) of lists/tables describing the selected objects for each layer.  This would apply to the Python and JavaScript APIs.

  -------------------------------------------------------------------------

# 3. aladin-lite interfaces

## 3.1. Listeners

Supported listeners that enable us to attach callback functions to
events that occur within aladin-lite.

  **Event Name**     |**Description**              |**Data**            |**Supported**
  ------------------ |---------------------------- |------------------- |---------------
  objectsSelected    |Fired when objects are selected by region, represents the individual objects selected |List of sources     |Existing
  regionsSelected    |Fired when objects are selected by region, represents the region used to select the objects    |Shape of the region selected|New
  objectClicked      |Fires when an object is clicked | Catalog object and xy coordinates of the click event | Existing
  footprintClicked   |Fires when a footprint is clicked | Footprint object xy coordinates of the click event | Existing
  -----------------------------------------------------------------------------------

## 3.2. GET

#### GetSelectedSources
```
/**
 * @returns List[Sources]
 */
function GetSelectedSources()
```
Returns the list of currently selected source objects.

> #### GetSelectedRegion

Reviewer: Related to the previous discussion about how regions are handled/maintained, maybe one possibility is that Aladin Lite doesn't maintain any persistence of regions, but just provides the building blocks for creating regions, leaving the management of persistent regions and applying them to selections to the Python layer.

Author: That's pretty much how the [demo implementation](https://github.com/pcuste1/aladin-lite/pull/1/files) I've provided functions! I added a "regionSelected" callback to Aaldin-lite that returns the building blocks for the region that was selected, and the [ipyaladin demo](https://github.com/pcuste1/ipyaladin/pull/3/files) uses those to generate astropy regions

```
/**
 * @returns Region
 */
function GetSelectedRegion()
```
Returns the region definition of the selection event, if a region was
used to select. 

#### GetSelectedFootprints
```
/**
 *@returns List[Footprint]
 */
function GetSelectedFootprints()
```
Returns the currently selected footprints

## 3.3. SELECT

#### SelectSources 

##### Option 1: 
```
/**
 * @param sources: List[Source]
 * @param keys: List[str]
 */

> function SelectSources(sources, keys)

Reviewer: The details of how this would look are not clear to me, and they may not be very important yet, but an example of how it might look could be helpful.

Author: Here is what an [example implementation](https://github.com/pcuste1/ipyaladin/pull/2) might look like. I can add a link to the doc as well

determine if a source should be selected

  **Parameter**   |**Type**    |**Description**
  --------------- |----------- |---------------------------------------------------
  match_func      |func(a,b)   |The function used to determine if an object should be selected

  -------------------------------------------------------------------------------

#### SelectRegion
```
/**
 * @param region: Region
 */
function SelectRegion(region)
```
Triggers a region selection within aladin-lite widget

  **Parameter**             |**Type**          |**Description**
  ------------------------- |----------------- |---------------------------
  region                    |Region            |The region to select objects within

  -----------------------------------------------------------------------

#### SelectFootprint
```
/**
 * @param footprint: Footprint
 */
function SelectFootprint(footprint)
```
Triggers a footprint selection within aladin-lite widget

  **Parameter**       |**Type**         | **Description**
  ------------------- |---------------- |-----------------------------------
  footprint           |Footprint        |The footprint to select

## 3.4. Misc

> #### RemoveSelection

Reviewer: Is this meant to remove all selections including footprints and sources with possible multiple layers?

Author: I intended for this to be a pass through to the [unselectObjects](https://github.com/cds-astro/aladin-lite/blob/master/src/js/View.js#L1463) method in aladin-lite. My understanding is that is unselects both footprints and sources
